### ISSUE TITLE
Addon attachments uses invalid addon identifier, forces replacement

### ISSUE THREAD

**Issue by robinvdvleuten** [heroku/terraform-provider-heroku#311]
### Terraform + Heroku Provider Version
Run `terraform -v` to show core and any provider versions. A sample output could be:

```
Terraform v0.15.3
on darwin_amd64
+ provider registry.terraform.io/heroku/heroku v4.4.1
```

### Affected Resource(s)
- heroku_addon_attachment

### Terraform Configuration Files
```hcl-terraform
resource "heroku_addon" "cache" {
  app  = heroku_app.api.name
  plan = "heroku-redis:hobby-dev"
}

resource "heroku_addon_attachment" "cache" {
  app_id   = heroku_app.api.id
  addon_id = heroku_addon.cache.id
  name     = "REDIS_CACHE"
}
```

### Output
```
  # heroku_addon_attachment.cache must be replaced
-/+ resource "heroku_addon_attachment" "cache" {
      ~ addon_id  = "756ae678-6302-49ef-9be7-fdf827b70fc3" -> "redis-transparent-46729" # forces replacement
      ~ id        = "662702bd-c05e-41b6-a8ff-09aa2edde56d" -> (known after apply)
        name      = "REDIS_CACHE"
      + namespace = (known after apply)
        # (1 unchanged attribute hidden)
    }
```

### Expected Behavior
I expect that the `addon_id` is either the name of the addon or the identifier.

### Actual Behavior
When I set it to its UUID (like mentioned in the docs), it tries to replace it with the name. Then when I run apply, it stores the UUID in the state. So Terraform tries to replace the attachment everytime it runs.

### Steps to Reproduce
1. `terraform apply`

**Comment by robinvdvleuten** [comment]
So while digging through the state, it seems that the name is stored for both the name and id attribute;

```json
{
  "mode": "managed",
  "type": "heroku_addon",
  "name": "cache",
  "provider": "provider[\"registry.terraform.io/heroku/heroku\"]",
  "instances": [
    {
      "schema_version": 2,
      "attributes": {
        ...
        "id": "redis-transparent-46729",
        "name": "redis-transparent-46729",
        ...
      },
      ...
    }
  ]
}
```

**Comment by davidji99** [comment]
Hi @robinvdvleuten,

Is this a recently created terraform configuration/state? I ask as it's odd that your state file is showing the addon name for the `id` ("redis-transparent-46729") but it shows the UUID for my resource created two days ago:

```json
      "mode": "managed",
      "type": "heroku_addon",
      "name": "kafka",
      "provider": "provider[\"registry.terraform.io/heroku/heroku\"]",
      "instances": [
        {
          "schema_version": 2,
          ...
            "id": "cbe99122-779c-494a-a0e6-7f38a82b40ef", (not real UUID)
            "name": "kafka-rigid-14299",
            "plan": "heroku-kafka:private-standard-0",
          },
          "sensitive_attributes": [],
       }
     ]
```

What's more odd is while there is a state migration on the `heroku_addon` resource, it appears to have run for your resource given this: `"schema_version": 2,`.

Are all your `heroku_addon` resources in state showing `id` as the name?

**Comment by robinvdvleuten** [comment]
I have created the project just yesterday and the name is used for both the addons. But I created the Heroku resources manually a couple of days ago and was just migrating everything to this new Terraform project. So maybe something went wrong while importing the existing addons? When I change the ID to the correct UUIDs manually in the state, Terraform works correctly.

**Comment by davidji99** [comment]
Can you paste the import commands you used if possible?

I suspect that you might have done `terraform import heroku_addon.cache redis-transparent-46729`, which is totally valid. If so, I see a potential improvement on the resource's import behavior and documentation.

**Comment by robinvdvleuten** [comment]
What I did was `terraform import heroku_addon_attachment.cache ddb72bf1-194e-4e4c-9d1f-2e5ef24e5193`, because when I use the name while importing it gives an error that the UUID must be used.

**Comment by davidji99** [comment]
Oh, you imported the addon attachment but not the addon? That still doesn't explain how your state for the addon ended up like this:

```
{
  "mode": "managed",
  "type": "heroku_addon",
  "name": "cache",
  "provider": "provider[\"registry.terraform.io/heroku/heroku\"]",
  "instances": [
    {
      "schema_version": 2,
      "attributes": {
        ...
        "id": "redis-transparent-46729",
        "name": "redis-transparent-46729",
        ...
      },
      ...
    }
  ]
}
```

**Comment by robinvdvleuten** [comment]
I created the terraform project and then imported all existing resources (including both the addon and the addon attachment).

**Comment by davidji99** [comment]
Cool, can you tell me the command you used to import the addon?

**Comment by robinvdvleuten** [comment]
I've used this command to import the addon and its attachment;

```
terraform import heroku_addon.cache <UUID of addon>
terraform import heroku_addon_attachment.cache <UUID of addon attachment>
```

**Comment by devopsix** [comment]
We are having the exact same issue with attaching a PostgreSQL database to other dynos. Affects both addons and attachments created through Terraform as well as imported ones.

![image](https://user-images.githubusercontent.com/58369032/152173018-32e13934-3f06-4302-bdd1-c31f7b0aa49b.png)

In the state, the addon attachment refers to the addon by its UUID.

![image](https://user-images.githubusercontent.com/58369032/152173977-db67d302-b78e-4f38-8421-1d210fd1a577.png)

Addon ID and name look OK:

![image](https://user-images.githubusercontent.com/58369032/152174332-63c3d53a-febc-4a34-88fc-a0ff6fd60d93.png)

**Comment by mars** [comment]
So, I just opened a [fresh draft PR](https://github.com/heroku/terraform-provider-heroku/pull/332) 🍃 to fix this by changing all resources to use app IDs. It's a breaking, major-version changeset that would finally fix a variety of these `id` 🆚 `name` problems by standardizing this provider to app ID.

Would love to hear perspectives of how to move this forward on that PR 🙏

**Comment by devopsix** [comment]
Thanks a lot for providing an update on this!

We have tried version 5.0.0 of the provider with some test infrastructure. Unfortunately, for us it does not seem to work as we would expect. We are hesitating using 5.0.0 for production infrastructure as we see a potentially very harmful impact on databases addons managed through the provider.

### Scenario 1: Upgrading infrastructure originally created with provider 4.8.0

Previously, our configuration looked like this:


```tf
resource "heroku_addon" "database" {
  depends_on = [module.backend]
  app = var.backend_name
  plan = "heroku-postgresql:${var.backend_database_plan}"
}

data "heroku_app" "backend-tasks-worker" {
  depends_on = [module.backend-tasks-worker]
  name = var.backend_tasks_worker_name
}

resource "heroku_addon_attachment" "backend-tasks-worker-database-backend" {
  name = "DATABASE"
  app_id = data.heroku_app.backend-tasks-worker.id
  addon_id = heroku_addon.database.id
}
```

I.e. we were already using `heroku_addon_attachment.app_id` but we were using `heroku_addon.app` with the app's name.

Running `terraform apply` with version 4.8.0 of the provider:

- Does not alter the database addon
- Does not detach/re-attach the database from/to the owning dyno
- Replaces the `heroku_addon_attachment` resource and hence detaches/re-attaches the database from/to the other dyno every time

![image](https://user-images.githubusercontent.com/58369032/154944988-317c2141-c97d-4567-beee-e138d4cc7710.png)

![image](https://user-images.githubusercontent.com/58369032/154945207-7db6770f-e506-4345-a67e-ceb77da7143d.png)

After upgrading to 5.0.0 and following the [migration guide](https://registry.terraform.io/providers/heroku/heroku/latest/docs/guides/upgrading), our configuration looks like this:

```tf
# NEW
data "heroku_app" "backend" {
  depends_on = [module.backend]
  name = var.backend_name
}

resource "heroku_addon" "database" {
  depends_on = [module.backend]
  # CHANGED app => app_id (as per required configuration change #1)
  # CHANGED var.backend_name (equivalent to data.heroku_app.backend.name) => data.heroku_app.backend.id (as per required configuration change #2)
  app_id = data.heroku_app.backend.id
  plan = "heroku-postgresql:${var.backend_database_plan}"
}

data "heroku_app" "backend-tasks-worker" {
  depends_on = [module.backend-tasks-worker]
  name = var.backend_tasks_worker_name
}

resource "heroku_addon_attachment" "backend-tasks-worker-database-backend" {
  name = "DATABASE"
  app_id = data.heroku_app.backend-tasks-worker.id
  addon_id = heroku_addon.database.id
}
```

We have to admit, we are not entirely sure how to understand [required change #2](https://registry.terraform.io/providers/heroku/heroku/latest/docs/guides/upgrading#2-replace-heroku_app-data-source-result-references-with-app_id). Neither in 4.8.0 nor in 5.0.0 the `heroku_app` data source has an `app` or `app_id` attribute. We read this to describe the need to change `heroku_app.example.name` into `heroku_app.example.id`.

When running `terraform apply` for the first time, it:

- Detects external changes for the `heroku_addon` and `heroku_addon_attachment` resources (Expected as per the migration guide)
- Replaces the `heroku_addon` database resource (Evil: this kills our database!)
- Replaces the `heroku_addon_attachment` resource (the original issue)

![image](https://user-images.githubusercontent.com/58369032/154951669-e57dbd57-4e5a-429e-a7a9-f56903aed63a.png)
![image](https://user-images.githubusercontent.com/58369032/154952053-9ca8e5ab-4f7a-422e-aae8-1f2e9d43c9f1.png)

We are still investigating if some extra migration step would be necessary, e.g. manually updating Terraform state. (We may possibly be misreading [required configuration change #2](https://registry.terraform.io/providers/heroku/heroku/latest/docs/guides/upgrading#2-replace-heroku_app-data-source-result-references-with-app_id)).

Every time running `terraform apply` thereafter, it:

- Detects no further external changes for the `heroku_addon` and `heroku_addon_attachment` resources
- But keeps replacing the `heroku_addon` database resource
- And keeps replacing the `heroku_addon_attachment` resource

![image](https://user-images.githubusercontent.com/58369032/154954000-207a53f0-55c0-4138-a8df-ca13a5092666.png)

The dyno owning the database:

![image](https://user-images.githubusercontent.com/58369032/154955806-19e81096-f4ab-49be-9767-dee8ba0b80f6.png)

The other dyno:

![image](https://user-images.githubusercontent.com/58369032/154956061-b266aaa6-b2db-4f53-9834-9d45e8bd1f4e.png)

This may or may no be related: Terraform keeps detecting external changes to both the apps, the one owning the database addon and the one the database is attached to additionally. This can maybe be explained by the config variable for the database URL now changing every time.

![image](https://user-images.githubusercontent.com/58369032/154954247-7eb67ccb-5552-4068-8f90-439fcdbb44e2.png)

### Scenario 2: Creating new infrastructure with provider 5.0.0

Using the same configuration as shown above for “after upgrading to 5.0.0”, creating new infrastructure from scratch works as expected. But for every subsequent run of `terraform apply` we are experiencing the exact same issue as in the “upgrade from 4.8.0” scenario

**Comment by mars** [comment]
Sorry this didn't fix the problem. Re-opening this issue. I will try to to carve out time later this week to re-investigate.

Regarding the Upgrade step #2: the `heroku_app` data source did have an undocumented `app` attribute. That attribute was migrated to `app_id`, and is now documented. If you weren't using that data source attribute, then no further action is required.

**Comment by mars** [comment]
Ah, I see the `heroku_app` data source app ID is incorrectly documented in the Upgrade Guide. Should be `id` not `app_id`.

**Comment by devopsix** [comment]
Thank you very much for your prompt response, @mars ! In the meantime I have updated my comment as we found that the issue we experienced is related to use of `depends_on` with data sources.

**Comment by mars** [comment]
Okay @devopsix, I'm re-closing this, as I know this issue's original problem was solved by the `v5.0.0` **Apps by ID** update.

Yes, `depends_on` with data sources is very tricky. [Their lifecycle](https://www.terraform.io/language/data-sources#data-source-lifecycle) then defers the data query to the `apply` phase. This behavior has also changed over Terraforms releases: back in v0.12 and earlier `depends_on` would not work at all for data sources.

This is frankly abusing the data sources, because data sources are supposed to represent data from resources that are manage outside of the Terraform configuration, not dynamically lookup resources created/managed within the same configuration.

The [docs are clear about this](https://www.terraform.io/language/data-sources):
> Data sources allow Terraform to use information defined outside of Terraform, defined by another separate Terraform configuration, or modified by functions.


I'd recommend making the Terraform module output the App ID values directly. The best way to get a resource attribute is to pass it through variables, not look it up again, separately.

**Comment by mars** [comment]
Thanks @devopsix for pointing out the confusing Upgrading docs. We've [corrected that section](https://registry.terraform.io/providers/heroku/heroku/latest/docs/guides/upgrading#2-replace-heroku_app-data-source-result-references-with-id).
